hardware_interface/src/mock_components/generic_system.cpp

```diff
@@ -369,7 +382,124 @@ std::vector<hardware_interface::CommandInterface> GenericSystem::export_command_
+      }
+
+      if (key == info_.joints[joint_index].name + "/" + hardware_interface::HW_IF_POSITION)
+      {
+        joint_found_in_x_requests_[joint_index] += 1;
+      }
+      if (key == info_.joints[joint_index].name + "/" + hardware_interface::HW_IF_VELOCITY)
+      {
+        joint_found_in_x_requests_[joint_index] += 1;
+      }
+      if (key == info_.joints[joint_index].name + "/" + hardware_interface::HW_IF_ACCELERATION)
+      {
+        joint_found_in_x_requests_[joint_index] += 1;
+      }
+    }
+    else
+    {
+      RCUTILS_LOG_DEBUG_NAMED(
+        "mock_generic_system", "Got interface '%s' that is not joint - nothing to do!",
+        key.c_str());
+    }
+  }
+
+  for (size_t i = 0; i < info_.joints.size(); ++i)
+  {
+    // There has to always be at least one control mode from the above three set
+    if (joint_found_in_x_requests_[i] == FOUND_ONCE_FLAG)
+    {
+      RCUTILS_LOG_ERROR_NAMED(
+        "mock_generic_system", "Joint '%s' has to have '%s', '%s', or '%s' interface!",
+        info_.joints[i].name.c_str(), hardware_interface::HW_IF_POSITION,
+        hardware_interface::HW_IF_VELOCITY, hardware_interface::HW_IF_ACCELERATION);
+      ret_val = hardware_interface::return_type::ERROR;
+    }
+
+    // Currently we don't support multiple interface request
+    if (joint_found_in_x_requests_[i] > (FOUND_ONCE_FLAG + 1))
+    {
+      RCUTILS_LOG_ERROR_NAMED(
+        "mock_generic_system",
+        "Got multiple (%zu) starting interfaces for joint '%s' - this is not "
+        "supported!",
+        joint_found_in_x_requests_[i] - FOUND_ONCE_FLAG, info_.joints[i].name.c_str());
+      ret_val = hardware_interface::return_type::ERROR;
+    }
+  }
+
+  return ret_val;
+}
+
+return_type GenericSystem::perform_command_mode_switch(
+  const std::vector<std::string> & start_interfaces,
+  const std::vector<std::string> & /*stop_interfaces*/)
+{
+  if (!calculate_dynamics_)
+  {
+    return hardware_interface::return_type::OK;
+  }
+
+  for (const auto & key : start_interfaces)
+  {
+    // check if interface is joint
+    auto joint_it_found = std::find_if(
+      info_.joints.begin(), info_.joints.end(),
+      [key](const auto & joint) { return (key.find(joint.name) != std::string::npos); });
+
+    if (joint_it_found != info_.joints.end())
+    {
+      const size_t joint_index = std::distance(info_.joints.begin(), joint_it_found);
+
+      if (key == info_.joints[joint_index].name + "/" + hardware_interface::HW_IF_POSITION)
+      {
+        joint_control_mode_[joint_index] = POSITION_INTERFACE_INDEX;
+      }
+      if (key == info_.joints[joint_index].name + "/" + hardware_interface::HW_IF_VELOCITY)
+      {
+        joint_control_mode_[joint_index] = VELOCITY_INTERFACE_INDEX;
+      }
+      if (key == info_.joints[joint_index].name + "/" + hardware_interface::HW_IF_ACCELERATION)
+      {
+        joint_control_mode_[joint_index] = ACCELERATION_INTERFACE_INDEX;
+      }
+    }
+  }
```

Comment:
```suggestion
  for (const auto & key : start_interfaces)
  {
    // check if interface is joint
    auto joint_it_found = std::find_if(
      info_.joints.begin(), info_.joints.end(),
      [key](const auto & joint) { return (key.find(joint.name) != std::string::npos); });

    if (joint_it_found != info_.joints.end())
    {
      const size_t joint_index = std::distance(info_.joints.begin(), joint_it_found);

      if (key == info_.joints[joint_index].name + "/" + hardware_interface::HW_IF_POSITION)
      {
        joint_control_mode_[joint_index] = POSITION_INTERFACE_INDEX;
      }
      if (key == info_.joints[joint_index].name + "/" + hardware_interface::HW_IF_VELOCITY)
      {
                if (!calculate_dynamics_)
        {
          RCUTILS_LOG_WARN_NAMED(
            "mock_generic_system",
            "Requested velocity mode for joint '%s' without dynamics calculation enabled. Without "
            "dynamic "
            "calculation only position interfaces are supported.",
            info_.joints[joint_index].name.c_str());
        }
        else
        {
          joint_control_mode_[joint_index] = VELOCITY_INTERFACE_INDEX;
        }
      }
      if (key == info_.joints[joint_index].name + "/" + hardware_interface::HW_IF_ACCELERATION)
      {
        if (!calculate_dynamics_)
        {
          RCUTILS_LOG_WARN_NAMED(
            "mock_generic_system",
            "Requested acceleration mode for joint '%s' without dynamics calculation enabled. "
            "Without dynamic "
            "calculation only position interfaces are supported.",
            info_.joints[joint_index].name.c_str());
        }
        else
        {
          joint_control_mode_[joint_index] = ACCELERATION_INTERFACE_INDEX;
        }
      }
    }
  }
```
I think it would be good to warn people that this is actually not possible. Maybe even returning ERROR would be an option here?

On a side node: I tested with returning ERROR and the result was that the velocity-based controller was in state active afterwards, which seems wrong.

Reply:
@fmauch I have moved this into "prepare_for_command_mode_switch" method. There is a better place for this - becasue the method is non-realitme. When you return error from that method then the controller will not be activated. We are currently expecting that "perform_command_mode_switch" works. You are right, we need to add more logic there.